java-checks/src/main/java/org/sonar/java/checks/serialization/RecordSerializationIgnoredMembersCheck.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.serialization;
+
+import java.util.Collections;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S6209")
+public class RecordSerializationIgnoredMembersCheck extends IssuableSubscriptionVisitor {
+
+  private static final MethodMatchers SERIALIZABLE_MATCHERS = MethodMatchers.or(
+    methodMatcher("readObjectNoData"),
+    methodMatcher("readObject", "java.io.ObjectInputStream"),
+    methodMatcher("readExternal", "java.io.ObjectInput"),
+    methodMatcher("writeObject", "java.io.ObjectOutputStream"),
+    methodMatcher("writeExternal", "java.io.ObjectOutput"));
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.RECORD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    for (Tree member : ((ClassTree) tree).members()) {
+      if (member.is(Tree.Kind.VARIABLE)) {
+        checkField((VariableTree) member);
+      } else if (member.is(Tree.Kind.METHOD)) {
```

Comment:
To increase precision, we could even check that the `Serializable` related methods are private and that the `Externalizable` methods are overriding. But that could also be tackled as an improvement to the rule

Reply:
Since the method signatures are matching, I'm not sure this is required, and I feel it's OK to let it that way.